src/builder.rs

```diff
@@ -657,6 +667,39 @@ impl<V> Bundle<InProgress<Proof, PartiallyAuthorized>, V> {
     }
 }
 
+/// A trait that provides a minimized view of an Orchard input suitable for use in
+/// fee and change calculation.
+pub trait InputView<NoteRef> {
+    /// An identifier for the input being spent.
+    fn note_id(&self) -> &NoteRef;
+    /// The value of the input being spent.
+    fn value<V: From<u64>>(&self) -> V;
+}
+
+impl InputView<()> for SpendInfo {
+    fn note_id(&self) -> &() {
+        // The builder does not make use of note identifiers, so we can just return the unit value.
```

Comment:
Could this then be an uninhabited type like `std::convert::Infallible`, such that this method will panic if called?

Reply:
I'm dubious about this API because `InputView<anything>` does not seem to be the right type. We discussed this in PR gardening and @str4d or @nuttycom were going to open an issue.

Reply:
I followed how the API was used [here](https://github.com/zcash/librustzcash/blob/903922c59ac85ce56aa0e13a8a62e7128f5c555e/zcash_primitives/src/transaction/components/sapling/builder.rs#L72-L82) without giving it a significant amount of thought. I don't really understand the use-case for note_id, although I see that it's in use in `zcash_client_backend`. I'm concerned about having an explicit panic, if it comes up where someone wants to be generic for `T: InputView`, a method that might noop is a lot easier to work with than a method that might panic.

Reply:
The panic is actually safe, because if someone is trying to do something in their generic code with the uninhabited return type of that method, the compiler will prevent you from using that operation. This is the great value of parametricity; we can statically exclude that the method will be called in situations where it would be unsafe to do so.

Reply:
I'm not sure what you mean. From a quick [playground test](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=406eab482094d7ce70881e4172e7c15e), when you actually call the generic code with the type determined, the compiler knows it's a panic, and will give unused code warnings about any code after the panic, but the code compiles fine and panics at runtime. If you comment out line 5, the only compiler warning is an unused variable. There are no compiler warnings associated with the `foo` fn itself. 

It's not a method zingolabs is likely to use, and the compiler does seem to do a reasonable job of warning consumers about panics.. The only case I can think of where this could realistically be missed (without ignoring compiler warnings) would be a library that exposes a generic consumer of `InputView` without having any calls to that function that use the specific concrete type in question. This scenario might be too contrived to bother worrying about. 

If you agree this is too contrived to be worth worrying about, I'm happy to swap out the unit type for an uninhabited type.

Reply:
@nuttycom If this is waiting on me, let me know. I'm not convinced on uninhabited-vs-unit type, but the downside feels a bit contrived so I'm happy to be overruled here.

Reply:
My rationale here is that if one is using `InputView` as a constraint, this could potentially cause unexpected behavior if, for example, the note ID is used as a key in a map. Making it panic on an invalid call is safer from that perspective.

Reply:
I just realized though that this is a clone of what was done for Sapling, so this is fine.